### PR TITLE
[WIP] Make Index.values read-only

### LINKED ISSUE #33001: Idea: Make pd.Index.values not writable
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

a = pd.DataFrame({"col1": [1, 2, 3]}, index=["a", "b", "c"])
b = a.copy()

b.index.values[0] = "newval"
assert a.index[0] == "newval"
```
#### Problem description

`Index` objects are assumed to be immutable, but wrap a mutable object. In my experience, it's pretty common to just grab the `.values` from pandas object. This has caused surprising behaviour in my own code, and has been discussed here before https://github.com/pandas-dev/pandas/issues/19862.

I would like to make sure that the values numpy array for index objects has it's`WRITABLE` flag set to false, to make `Index` object more immutable.

Does this sound reasonable? If so, I try at a PR.

#### Expected Output

```python
assert a.index[0] != "newval"
```

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.6.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.3.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.0.3
numpy            : 1.18.2
pytz             : 2018.9
dateutil         : 2.7.5
pip              : 20.0.2
setuptools       : 46.0.0
Cython           : 0.29.14
pytest           : 5.3.5
hypothesis       : 5.6.0
sphinx           : 2.4.4
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.4.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.1
IPython          : 7.13.0
pandas_datareader: None
bs4              : 4.8.2
bottleneck       : None
fastparquet      : 0.3.3
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : 3.2.0
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 2.6.2
pandas_gbq       : None
pyarrow          : 0.16.0
pytables         : None
pytest           : 5.3.5
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : 3.6.1
tabulate         : 0.8.3
xarray           : 0.15.0
xlrd             : 1.2.0
xlwt             : None
xlsxwriter       : None
numba            : 0.48.0

</details>

### PR DESCRIPTION
**Pull request by ivirshup**
- [ ] closes #33001
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

This is a PR for looking around enforcing immutability of `pd.Index` objects. It will be a draft, but the main idea is to show what tests fail and what would need to change for this to be implemented. Because of this, I'm going to add some commits incrementally, so there is a history in the build logs.

### CODE REVIEW

**Comment by alimcmaster1** [comment]
@ivirshup are you still interested in working on this?

**Comment by ivirshup** [comment]
I've gotten a little stuck on the cython part. Using the cython 3.0 alpha, the errors about using a `const` fused type go away, but are replaced by parse errors of the generated c code. This seems to only be the case if the types are imported from numpy.

I think it could make sense for someone who has more experience with cython to take over, or at least figure out that part. I'm still interested, but I'm not sure I have the time to learn that much about cython at the moment.

**Comment by jreback** [comment]
see if you can merge master and get this passing

File: pandas/tests/indexes/test_any_index.py [hunk 1]
```diff
@@ -33,6 +33,20 @@ def test_mutability(index):
         index[0] = index[0]
 
 
+def test_values_mutability(index):
+    if not len(index):
+        return
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/test_any_index.py:38]: moot point till tests are passing, but ideally should use pytest.skip with message.

**Comment by ivirshup** [comment]
@jreback, IIRC this requires cython 3. Since master can't compile with cython 3 (https://github.com/pandas-dev/pandas/issues/34014), I think this is stuck.

**Comment by jbrockmendel** [comment]
closing as inactive, we can reopen if you like after cy3 is available.

**Pull request closed without merging**